### ISSUE TITLE
Incorrect error "Expected 1 argument, 0 found"

### ISSUE THREAD

**Issue by djc** [rust-analyzer/rust-analyzer#5502]
I'm on the nightly channel, using VS Code. Since yesterday, I'm seeing this issue:

![Screen Shot 2020-07-23 at 11 18 33](https://user-images.githubusercontent.com/158471/88270658-41f71180-ccd6-11ea-962a-d14732537da7.png)

The `row` in question is a [`Row`](https://docs.rs/tokio-postgres/0.5.5/tokio_postgres/row/struct.Row.html). However, `cargo c` on the command line thinks my code is fine.

**Comment by bjorn3** [comment]
Is the type of `row` correctly inferred? If not this is probably https://github.com/rust-analyzer/rust-analyzer/issues/1791. If the type is not known rust-analyzer guesses what method you meant. In this case it guessed wrong. Before https://github.com/rust-analyzer/rust-analyzer/pull/5283 argument count mismatches were not reported, so there was no error shown.

**Comment by djc** [comment]
Yeah, the type seems to be correctly inferred, see:

![Screen Shot 2020-07-23 at 11 33 38](https://user-images.githubusercontent.com/158471/88272025-5cca8580-ccd8-11ea-988f-9e38f63850c6.png)

**Comment by djc** [comment]
Calls to `row.get()` in this crate:

```
backend/src/documents.rs
75:    let ctype: String = row.get(0);
76:    let bytes: Vec<u8> = row.get(1);

backend/src/assets.rs
58:        let maturity = row.get::<_, NaiveDate>(4);
61:            id: row.get(0),
62:            industry: row.get(2),
63:            sector: row.get(3),
65:            interest_rate_margin: row.get(5),
66:            rating: row.get(6),
67:            esg: (row.get(7), row.get(8)),
68:            level: row.get::<_, Option<Access>>(9),
69:            auditor: row.get(10),
72:        match row.get(1) {
116:    let level = match row.get::<_, Option<Access>>(0) {
127:        name: row.get(0),
141:            id: row.get(0),
142:            name: row.get(1),
143:            uploader: row.get(2),
144:            uploaded: row.get(3),
147:        match row.get(4) {
160:        .map(|row| row.get(0))?;

backend/src/authn.rs
47:        Some(row) => (row.get(0), row.get(1), row.get(2)),
135:            Some(row) => row.get(0),
163:            Some(row) => (row.get(0), row.get(1)),
208:            row.get::<_, &str>(0),

backend/src/lib.rs
62:                .map(|row| row.get::<_, String>(0))
```

Looks like the reason this is different might be that a non-static lifetime is involved?

**Comment by flodiebold** [comment]
We ignore lifetimes completely, that's probably not the problem.

Am I reading that right, it says "found 0" even though there is an argument? That's weird.

Also, can you check what the method resolved to?

**Comment by djc** [comment]
The method also seems to resolve correctly:

![Screen Shot 2020-07-23 at 11 58 43](https://user-images.githubusercontent.com/158471/88274356-de6fe280-ccdb-11ea-9f8a-958b595610bf.png)

**Comment by jonas-schievink** [comment]
That span looks very odd. It should encompass the entire call expression but seems to stop inside the generic arguments. Would it be possible to get a minimal reproduction of this issue?

**Comment by djc** [comment]
This is not enough to reproduce:

```rust
use tokio_postgres::{Error, NoTls};

#[tokio::main]
async fn main() -> Result<(), Error> {
    let (client, connection) =
        tokio_postgres::connect("host=localhost user=postgres", NoTls).await?;
    tokio::spawn(async move { connection.await.unwrap() });
    let row = client
        .query_one("SELECT $1::TEXT", &[&"hello world"])
        .await?;

    let foo = "hello".to_string();
    let s = format!(
        "https://this.will.make.for.a.very.very.very.long/url/to/append/to/{}/and/more/{}",
        row.get::<_, &str>(0),
        foo,
    );

    bar(&s);
    Ok(())
}

fn bar(_: &str) -> bool {
    true
}
```

One other interesting thing I noticed: there seems to be a caching issue. If I reload my VS Code window, the error does not show up immediately. Only if I make some edits that result in incorrect syntax it starts showing up, and then doesn't go away again if I correct such errors.

**Comment by jonas-schievink** [comment]
Hmm, this might not be a bug in the argument count mismatch diagnostic then

**Comment by djc** [comment]
I bisected this and it seems to have started with 63ce2c7b5fd96e6688796f2ddd1cd7316df8d11d.

**Comment by kjeremy** [comment]
Did it have 0 arguments at one time? I wonder if the diagnostics are stale.

**Comment by NikosEfthias** [comment]
I have the same issue with tokio postgress and the previous version of the rust analyzer seemed to work just fine

**Comment by jonas-schievink** [comment]
If this happened over the last couple of days, then maybe it was caused by https://github.com/rust-analyzer/rust-analyzer/pull/5481? Which version did this break in?

**Comment by kjeremy** [comment]
@jonas-schievink that's what I'm wondering...

**Comment by kjeremy** [comment]
My guess is that we need to grab the file version before the diagnostics start computing because we're probably sending old diagnostics for a newer file version.

**Comment by GopherJ** [comment]
![image](https://user-images.githubusercontent.com/33961674/88429304-add09b80-cdf6-11ea-90b6-9b66def0bf0c.png)
same here

**Comment by kjeremy** [comment]
If you change the version to `None` here does it go away?

https://github.com/rust-analyzer/rust-analyzer/blob/14a3a713c74108626fe356dad21d93b84792878a/crates/rust-analyzer/src/main_loop.rs#L318

**Comment by djc** [comment]
This does not fix the problem:

```diff
diff --git a/crates/rust-analyzer/src/main_loop.rs b/crates/rust-analyzer/src/main_loop.rs
index e95d4157c..2c7369ba3 100644
--- a/crates/rust-analyzer/src/main_loop.rs
+++ b/crates/rust-analyzer/src/main_loop.rs
@@ -315,7 +315,7 @@ impl GlobalState {
                     .unwrap_or_default();

                 self.send_notification::<lsp_types::notification::PublishDiagnostics>(
-                    lsp_types::PublishDiagnosticsParams { uri: url, diagnostics, version },
+                    lsp_types::PublishDiagnosticsParams { uri: url, diagnostics, version: None },
                 );
             }
         }
```

**Comment by SorteKanin** [comment]
Getting a similar thing:

![image](https://user-images.githubusercontent.com/9417568/88462793-47a74f80-ceae-11ea-801d-28dd11ce8230.png)

No idea why this would pop up around my use statements. `cargo check` reports no errors.

**Comment by billyrieger** [comment]
![image](https://user-images.githubusercontent.com/43874200/88584372-83155b80-d01f-11ea-9081-2bb0d986ecca.png)

I'm having the same problem with `packed_simd`.

**Comment by museun** [comment]
I've ran into this same issue. It gives the error in two places in the file if I call the function from within a macro (`println!()`). Once at the beginning of the file and once where the call happens.

```rust
#![cfg_attr(debug_assertions, allow(dead_code,))]
use std::{fmt::Debug, str::FromStr};

struct Foo {}

impl Foo {
    fn something<'a, T, M>(&self, input: &str, data: M) -> Result<T, T::Err>
    where
        T: FromStr + Debug,
        T::Err: Debug,
    {
        todo!()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_something() {
        let foo = Foo {};
        // let res = foo.something::<usize, _>("foo", ());
        eprintln!("{:#?}", foo.something::<usize, _>("foo", ()));
    }
}
```

![err](https://user-images.githubusercontent.com/6295545/88593110-0f7a4b00-d02d-11ea-82ca-a25115fe7e36.png)

**Comment by lnicola** [comment]
In the last update you can disable the warning by turning off `rust-analyzer.diagnostics.enableExperimental`.

**Comment by NikosEfthias** [comment]
@lnicola how and where exactly do i do that ? I am using vim-lsp am I supposed to set this in vim lsp settings or something ?

**Comment by NikosEfthias** [comment]
apparently you guys broke everything. It doesnt show error messages doesnt suggest auto completion or anything anymore. I had to switch back to rls altough it is muchslower it at least works

**Comment by GopherJ** [comment]
I ran into the same issue by running `.take` on `Option`

![image](https://user-images.githubusercontent.com/33961674/94410006-2e08d980-0177-11eb-8069-1f745983bb28.png)

**Comment by lnicola** [comment]
> @lnicola how and where exactly do i do that ? I am using vim-lsp am I supposed to set this in vim lsp settings or something ?

Yes, I think it's called `diagnostics/enableExperimental`.

**Comment by flodiebold** [comment]
@djc are you still encountering this?

**Comment by vorner** [comment]
I haven't seen this in a while, but I can't say when exactly it stopped for me.

**Comment by djc** [comment]
Also haven't seen it in a while, but I haven't been using tokio-postgres that much recently (I no longer have access to the codebase I originally saw this in).

**Comment by flodiebold** [comment]
Ok, I'll close this then.
